torch/utils/data/datapipes/iter/grouping.py

```diff
@@ -27,15 +27,32 @@ class ShardingFilterIterDataPipe(IterDataPipe):
     """
     def __init__(self, source_datapipe: IterDataPipe):
         self.source_datapipe = source_datapipe
+        self.sharding_groups: Dict[int, Any] = {}
         self.num_of_instances = 1
         self.instance_id = 0
 
     def is_shardable(self):
         return True
 
-    def apply_sharding(self, num_of_instances, instance_id):
-        self.num_of_instances = num_of_instances
-        self.instance_id = instance_id
+    def apply_sharding(self, num_of_instances, instance_id, sharding_group=None):
+        if sharding_group is not None:
+            if None in self.sharding_groups:
+                raise Exception("Already defined default sharding and now trying to use the sharding group. "
+                                "This will lead to unexpected behavior.")
+        else:
+            if len(self.sharding_groups) and not None in self.sharding_groups:
+                raise Exception("Already defined non default sharding groups and now trying to use default one. "
+                                "This will lead to unexpected behavior.")
```

Comment:
Question: Can you elaborate on what is the unexpected behavior here?

No need to change the message though.